File: spec/fixed_length_encoder_spec.rb
require 'fixed_length_encoder'

def suppress_warnings
  original_verbosity = $VERBOSE
  $VERBOSE = nil
  result = yield
  $VERBOSE = original_verbosity
  return result
end

describe FixedLengthEncoder do
  describe 'Invalid encodings' do
    before (:each) do
      suppress_warnings do
        @origianl = FixedLengthEncoder::ALPHABET
        FixedLengthEncoder::ALPHABET = '0123456789'
      end
    end

    after (:each) do
      suppress_warnings do
        FixedLengthEncoder::ALPHABET = @origianl
      end
    end

    it 'should error on non-integers' do
      expect { FixedLengthEncoder.encode('ERROR') }.to raise_error(ArgumentError)
    end

    it 'shouldn\'t encode values too big for message length' do
      expect { FixedLengthEncoder.encode(100, 2) }.to raise_error(ArgumentError)
    end

    it 'shouldn\'t encode negative values' do
      expect { FixedLengthEncoder.encode(-1, 2) }.to raise_error(ArgumentError)
    end

    it 'should error for non-strings' do
      expect { FixedLengthEncoder.decode(0) }.to raise_error(ArgumentError)
    end

    it 'should error for bad characters' do
      expect { FixedLengthEncoder.decode('^') }.to raise_error(ArgumentError)
    end
  end

  describe 'Valid encodings' do
    before (:each) do
      suppress_warnings do
        @origianl = FixedLengthEncoder::ALPHABET
        @encode_map = FixedLengthEncoder::ENCODE_MAP
        @decode_map = FixedLengthEncoder::DECODE_MAP
        FixedLengthEncoder::ALPHABET = '0123456789abcdefghijklmnopqrstuvwxyzABCDEFGHIJKLMNOPQRSTUVWXYZ'
        max = 62*62 - 1
        encode_map = (0..max).to_a.shuffle
        decode_map = []
        (0..max).each { |i| decode_map[encode_map[i]] = i }
        FixedLengthEncoder::ENCODE_MAP = encode_map
        FixedLengthEncoder::DECODE_MAP = decode_map
      end
    end

    after (:each) do
      suppress_warnings do
        FixedLengthEncoder::ALPHABET = @origianl
        FixedLengthEncoder::ENCODE_MAP = @encode_map
        FixedLengthEncoder::DECODE_MAP = @decode_map
      end
    end

    it 'should be a valid alphabet' do
      FixedLengthEncoder.isValidAlphabet().should be_true
    end

    it 'should be a valid map' do
      FixedLengthEncoder.isValidMap().should be_true
    end

    it 'should be reversible for the min value' do
      value = 0
      message = FixedLengthEncoder.encode(value)
      FixedLengthEncoder.decode(message).should eq(value)
    end

    it 'should be reversible for the max value' do
      value = (62**8)-1
      message = FixedLengthEncoder.encode(value)
      FixedLengthEncoder.decode(message).should eq(value)
    end

    it 'should be reversible for the middle value' do
      value = ((62**8)/2).floor
      message = FixedLengthEncoder.encode(value)
      FixedLengthEncoder.decode(message).should eq(value)
    end
  end

  describe 'Default encodings' do
    it 'should be a valid alphabet' do
      FixedLengthEncoder.isValidAlphabet().should be_true
    end

    it 'should be a valid map' do
      FixedLengthEncoder.isValidMap().should be_true
    end

    it 'should be reversible for the min value' do
      value = 0
      message = FixedLengthEncoder.encode(value)
      FixedLengthEncoder.decode(message).should eq(value)
    end

    it 'should be reversible for the max value' do
      value = (36**8)-1
      message = FixedLengthEncoder.encode(value)
      FixedLengthEncoder.decode(message).should eq(value)
    end

    it 'should be reversible for the middle value' do
      value = ((36**8)/2).floor
      message = FixedLengthEncoder.encode(value)
      FixedLengthEncoder.decode(message).should eq(value)
    end
  end
end

File: spec/fixed_length_encoder_spec.js
var FixedLengthEncoder = require('../lib/fixed_length_encoder.js');

describe('FixedLengthEncoder', () => {
  var fle;

  describe('Invalid encodings', () => {
    beforeEach(() => {
      fle = new FixedLengthEncoder();
      fle.ALPHABET = '0123456789';
    });

    it('should error on non-integers', () => {
      expect(() => fle.encode('ERROR')).toThrow();
    });

    it('shouldn\'t encode values too big for message length', () => {
      expect(() => fle.encode(100, 2)).toThrow();
    });

    it('shouldn\'t encode negative values', () => {
      expect(() => fle.encode(-1, 2)).toThrow();
    });

    it('should error for non-strings', () => {
      expect(() => fle.decode(0)).toThrow();
    });

    it('should error for bad characters', () => {
      expect(() => fle.decode('^')).toThrow();
    });
  });

  describe('Valid encodings', () => {
    function shuffle(a) {
      for(var i = 0; i < a.length; i++) {
        var r = Math.floor(Math.random() * (i + 1));
        var t = a[i];
        a[i] = a[r];
        a[r] = t;
      }
      return a;
    }

    beforeEach(() => {
      fle = new FixedLengthEncoder();
      fle.ALPHABET = '0123456789abcdefghijklmnopqrstuvwxyzABCDEFGHIJKLMNOPQRSTUVWXYZ';
      var max = 62*62 - 1;
      var encode_map = shuffle(Array(max + 1).fill().map((x,i) => i));
      var decode_map = [];
      encode_map.forEach((cv, i) => decode_map[cv] = i);
      fle.ENCODE_MAP = encode_map
      fle.DECODE_MAP = decode_map
    });

    it('should be a valid alphabet', () => {
      expect(fle.isValidAlphabet()).toBe(true);
    });

    it('should be a valid map', () => {
      expect(fle.isValidMap()).toBe(true);
    });

    it('should be reversible for the min value', () => {
      var value = 0;
      var message = fle.encode(value);
      expect(fle.decode(message)).toEqual(value);
    });

    it('should be reversible for the max value', () => {
      value = Math.pow(62, 8) - 1;
      message = fle.encode(value);
      expect(fle.decode(message)).toEqual(value);
    });

    it('should be reversible for the middle value', () => {
      value = Math.pow(62, 8) - 1;
      value = Math.floor(value / 2);
      message = fle.encode(value);
      expect(fle.decode(message)).toEqual(value);
    });
  });

  describe('Default encodings', () => {
    beforeEach(() => {
      fle = new FixedLengthEncoder();
    });

    it('should be a valid alphabet', () => {
      expect(fle.isValidAlphabet()).toBe(true);
    });

    it('should be a valid map', () => {
      expect(fle.isValidMap()).toBe(true);
    });

    it('should be reversible for the min value', () => {
      var value = 0;
      var message = fle.encode(value);
      expect(fle.decode(message)).toEqual(value);
    });

    it('should be reversible for the max value', () => {
      value = Math.pow(36, 8) - 1;
      message = fle.encode(value);
      expect(fle.decode(message)).toEqual(value);
    });

    it('should be reversible for the middle value', () => {
      value = Math.pow(36, 8) - 1;
      value = Math.floor(value / 2);
      message = fle.encode(value);
      expect(fle.decode(message)).toEqual(value);
    });
  });
});
